src/aws-cpp-sdk-core/source/http/crt/CRTHttpClient.cpp

```diff
@@ -0,0 +1,639 @@
+            Aws::StringStream ss;
+            ss << SchemeMapper::ToString(uri.GetScheme()) << "://" << uri.GetAuthority() << ":" << uri.GetPort();
+
+            return ss.str();
+        }
+
+        // The main purpose of this is to ensure there's exactly one connection manager per unique endpoint.
+        // To do so, we simply keep a hash table of the endpoint key (see ResolveConnectionPoolKey()), and
+        // put a connection manager for that endpoint as the value.
+        // This runs in multiple threads potentially so there's a lock around it.
+        std::shared_ptr<Crt::Http::HttpClientConnectionManager> CRTHttpClient::GetWithCreateConnectionManagerForRequest(const std::shared_ptr<HttpRequest>& request, const Crt::Http::HttpClientConnectionOptions& options) const
+        {
+            const auto connManagerRequestKey = ResolveConnectionPoolKey(request->GetUri());
+
+            std::lock_guard<std::mutex> locker(m_connectionPoolLock);
+
+            const auto& foundManager = m_connectionPools.find(connManagerRequestKey);
+
+            // We've already got one, return it.
+            if (foundManager != m_connectionPools.cend()) {
+                return foundManager->second;
+            }
+
+            // don't have a manager for this endpoint, so create one for it.
+            Crt::Http::HttpClientConnectionManagerOptions connectionManagerOptions;
+            connectionManagerOptions.ConnectionOptions = options;
+            connectionManagerOptions.MaxConnections = m_configuration.maxConnections;
+            connectionManagerOptions.EnableBlockingShutdown = true;
+            //TODO: need to bind out Monitoring options to handle the read timeout config value.
+            // once done, come back and use it to setup read timeouts.
+
+            auto connectionManager = Crt::Http::HttpClientConnectionManager::NewClientConnectionManager(connectionManagerOptions);
+
+            if (!connectionManager)
+            {
+                return nullptr;
+            }
+
+            // put it in the hash table and return it.
+            m_connectionPools.emplace(connManagerRequestKey, connectionManager);
+
+            return connectionManager;
+        }
+
+        Crt::Http::HttpClientConnectionOptions CRTHttpClient::CreateConnectionOptionsForRequest(const std::shared_ptr<HttpRequest>& request) const
+        {
+            // connection options are unique per request, this is mostly just connection-level configuration mapping.
+            Crt::Http::HttpClientConnectionOptions connectionOptions;
+            connectionOptions.HostName = request->GetUri().GetAuthority().c_str();
+            // TODO: come back and update this when we hook up the rate limiters.
+            connectionOptions.ManualWindowManagement = false;
+            connectionOptions.Port = request->GetUri().GetPort();
+            
+            if (m_context.has_value() && request->GetUri().GetScheme() == Scheme::HTTPS) 

```

Comment:
Can m_context not have a value? It's an `Optional<T>`, but the CrtHttpClient constructor ALWAYS tries to create it, and sets m_bad=true if it fails.

Are you just trying to delay construction of the TlsContext? If so, `unique_ptr<T>` might make your intent more clear than `Optional<T>`.

I worry about complexity/weirdness trying to validate something down here in the guts of things, when that situation should be impossible. I'd probably just making this `unique_ptr<T>`, and not bother checking its validity down here in the guts, with the assumption that our 1 public function MakeRequest() already checked m_bad, and if that didn't happen then crash-o-rama

Reply:
we actually always create a tls context and I forgot that fact. Will leave as a value type, since its always there anyways.